File: hylife/utilitis_PIC/__pyccel__/STRUPHY_pusher.f90
module STRUPHY_pusher

use mappings_analytical, only: mapping_8z3wk9_g_inv => g_inv
use mappings_analytical, only: mapping_8z3wk9_df_inv => df_inv

use core, only: linalg_f7anys_det => det
use core, only: linalg_f7anys_transpose => transpose
use core, only: linalg_f7anys_matrix_matrix => matrix_matrix
use core, only: linalg_f7anys_matrix_vector => matrix_vector
implicit none




contains

!........................................
subroutine pusher_step3(particles, dt, b_part, u_part, kind_map, &
      params_map)

  implicit none
  real(kind=8), intent(inout)  :: particles (0:,0:)
  real(kind=8), value  :: dt
  real(kind=8), intent(in)  :: b_part (0:,0:)
  real(kind=8), intent(in)  :: u_part (0:,0:)
  integer(kind=8), value  :: kind_map
  real(kind=8), intent(in)  :: params_map (0:)
  real(kind=8), allocatable  :: b (:) 
  real(kind=8), allocatable  :: u (:) 
  real(kind=8), allocatable  :: b_prod (:,:) 
  real(kind=8), allocatable  :: xi (:) 
  real(kind=8), allocatable  :: v (:) 
  real(kind=8), allocatable  :: dfinv (:,:) 
  real(kind=8), allocatable  :: dfinv_t (:,:) 
  real(kind=8), allocatable  :: ginv (:,:) 
  real(kind=8), allocatable  :: temp_mat1 (:,:) 
  real(kind=8), allocatable  :: temp_mat2 (:,:) 
  real(kind=8), allocatable  :: temp_vec (:) 
  integer(kind=8)  :: np  
  integer(kind=8), allocatable  :: components (:,:) 
  integer(kind=8)  :: ierr  
  integer(kind=8)  :: ip  
  integer(kind=8)  :: i  
  integer(kind=8)  :: j  







  allocate(b(0:2_8))
  allocate(u(0:2_8))


  allocate(b_prod(0:2_8, 0:2_8))
  b_prod = 0.0


  allocate(xi(0:2_8))
  allocate(v(0:2_8))


  allocate(dfinv(0:2_8, 0:2_8))
  allocate(dfinv_t(0:2_8, 0:2_8))
  allocate(ginv(0:2_8, 0:2_8))


  allocate(temp_mat1(0:2_8, 0:2_8))
  allocate(temp_mat2(0:2_8, 0:2_8))


  allocate(temp_vec(0:2_8))


  np = size(particles(:, 0_8),1)


  allocate(components(0:2_8, 0:2_8))


  components(0_8, 0_8) = 11_8
  components(0_8, 1_8) = 12_8
  components(0_8, 2_8) = 13_8
  components(1_8, 0_8) = 21_8
  components(1_8, 1_8) = 22_8
  components(1_8, 2_8) = 23_8
  components(2_8, 0_8) = 31_8
  components(2_8, 1_8) = 32_8
  components(2_8, 2_8) = 33_8


  !$omp parallel
  !$omp do private(ip, b, u, xi, v, i, j, dfinv, dfinv_t, ginv, temp_mat1, temp_&
      !$omp &mat2, temp_vec) firstprivate(b_prod)
  do ip = 0, np - 1_8, 1


    b(0_8) = b_part(ip, 0_8)
    b(1_8) = b_part(ip, 1_8)
    b(2_8) = b_part(ip, 2_8)


    u(0_8) = u_part(ip, 0_8)
    u(1_8) = u_part(ip, 1_8)
    u(2_8) = u_part(ip, 2_8)


    b_prod(0_8, 1_8) = -b(2_8)
    b_prod(0_8, 2_8) = b(1_8)


    b_prod(1_8, 0_8) = b(2_8)
    b_prod(1_8, 2_8) = -b(0_8)


    b_prod(2_8, 0_8) = -b(1_8)
    b_prod(2_8, 1_8) = b(0_8)


    xi = particles(ip, 0_8:2_8)
    v = particles(ip, 3_8:5_8)


    do i = 0, 2_8, 1
      do j = 0, 2_8, 1
        dfinv(i, j) = mapping_8z3wk9_df_inv(xi(0_8), xi(1_8), xi(2_8), &
      kind_map, params_map, components(i, j))


      end do

    end do

    call linalg_f7anys_transpose(dfinv, dfinv_t)


    do i = 0, 2_8, 1
      do j = 0, 2_8, 1
        ginv(i, j) = mapping_8z3wk9_g_inv(xi(0_8), xi(1_8), xi(2_8), &
      kind_map, params_map, components(i, j))


      end do

    end do

    call linalg_f7anys_matrix_matrix(dfinv_t, b_prod, temp_mat1)
    call linalg_f7anys_matrix_matrix(temp_mat1, ginv, temp_mat2)
    call linalg_f7anys_matrix_vector(temp_mat2, u, temp_vec)


    particles(ip, 3_8) = dt*temp_vec(0_8) + particles(ip, 3_8)
    particles(ip, 4_8) = dt*temp_vec(1_8) + particles(ip, 4_8)
    particles(ip, 5_8) = dt*temp_vec(2_8) + particles(ip, 5_8)




  end do

  !evaluate inverse Jacobian matrix
  !transpose of inverse Jacobian matrix
  !evaluate inverse metric tensor
  !perform matrix-matrix and matrix-vector products
  !update particle velocities
  !$omp end do  
  !$omp end parallel  
  ierr = 0_8
end subroutine
!........................................

!........................................
subroutine pusher_step4(particles, dt, kind_map, params_map) 

  implicit none
  real(kind=8), intent(inout)  :: particles (0:,0:)
  real(kind=8), value  :: dt
  integer(kind=8), value  :: kind_map
  real(kind=8), intent(in)  :: params_map (0:)
  real(kind=8), allocatable  :: xi (:) 
  real(kind=8), allocatable  :: v (:) 
  real(kind=8), allocatable  :: dfinv (:,:) 
  integer(kind=8)  :: np  
  integer(kind=8), allocatable  :: components (:,:) 
  real(kind=8), allocatable  :: k1 (:) 
  real(kind=8), allocatable  :: k2 (:) 
  real(kind=8), allocatable  :: k3 (:) 
  real(kind=8), allocatable  :: k4 (:) 
  integer(kind=8)  :: ierr  
  integer(kind=8)  :: ip  
  integer(kind=8)  :: i  
  integer(kind=8)  :: j  






  allocate(xi(0:2_8))
  allocate(v(0:2_8))


  allocate(dfinv(0:2_8, 0:2_8))


  np = size(particles(:, 0_8),1)


  allocate(components(0:2_8, 0:2_8))


  components(0_8, 0_8) = 11_8
  components(0_8, 1_8) = 12_8
  components(0_8, 2_8) = 13_8
  components(1_8, 0_8) = 21_8
  components(1_8, 1_8) = 22_8
  components(1_8, 2_8) = 23_8
  components(2_8, 0_8) = 31_8
  components(2_8, 1_8) = 32_8
  components(2_8, 2_8) = 33_8


  allocate(k1(0:2_8))
  allocate(k2(0:2_8))
  allocate(k3(0:2_8))
  allocate(k4(0:2_8))


  !$omp parallel
  !$omp do private(ip, xi, v, dfinv, k1, k2, k3, k4)
  do ip = 0, np - 1_8, 1


    xi = particles(ip, 0_8:2_8)
    v = particles(ip, 3_8:5_8)


    do i = 0, 2_8, 1
      do j = 0, 2_8, 1
        dfinv(i, j) = mapping_8z3wk9_df_inv(xi(0_8), xi(1_8), xi(2_8), &
      kind_map, params_map, components(i, j))


      end do

    end do

    call linalg_f7anys_matrix_vector(dfinv, v, k1)


    do i = 0, 2_8, 1
      do j = 0, 2_8, 1
        dfinv(i, j) = mapping_8z3wk9_df_inv(dt*(k1(0_8)/Real(2_8, 8)) + &
      xi(0_8), dt*(k1(1_8)/Real(2_8, 8)) + xi(1_8), dt*(k1(2_8)/Real( &
      2_8, 8)) + xi(2_8), kind_map, params_map, components(i, j))


      end do

    end do

    call linalg_f7anys_matrix_vector(dfinv, v, k2)


    do i = 0, 2_8, 1
      do j = 0, 2_8, 1
        dfinv(i, j) = mapping_8z3wk9_df_inv(dt*(k2(0_8)/Real(2_8, 8)) + &
      xi(0_8), dt*(k2(1_8)/Real(2_8, 8)) + xi(1_8), dt*(k2(2_8)/Real( &
      2_8, 8)) + xi(2_8), kind_map, params_map, components(i, j))


      end do

    end do

    call linalg_f7anys_matrix_vector(dfinv, v, k3)


    do i = 0, 2_8, 1
      do j = 0, 2_8, 1
        dfinv(i, j) = mapping_8z3wk9_df_inv(dt*k3(0_8) + xi(0_8), dt*k3( &
      1_8) + xi(1_8), dt*k3(2_8) + xi(2_8), kind_map, params_map, &
      components(i, j))


      end do

    end do

    call linalg_f7anys_matrix_vector(dfinv, v, k4)


    particles(ip, 0_8) = modulo(0.166666666666667d0*dt*(2_8*k3(0_8) + k4 &
      (0_8) + 2_8*k2(0_8) + k1(0_8)) + 1.0d0*xi(0_8),1.0d0)
    particles(ip, 1_8) = modulo(0.166666666666667d0*dt*(2_8*k3(1_8) + k4 &
      (1_8) + 2_8*k2(1_8) + k1(1_8)) + 1.0d0*xi(1_8),1.0d0)
    particles(ip, 2_8) = modulo(0.166666666666667d0*dt*(2_8*k3(2_8) + k4 &
      (2_8) + 2_8*k2(2_8) + k1(2_8)) + 1.0d0*xi(2_8),1.0d0)




  end do

  !step 1 in Runge-Kutta method
  !step 2 in Runge-Kutta method
  !step 3 in Runge-Kutta method
  !step 4 in Runge-Kutta method
  !update logical coordinates
  !$omp end do  
  !$omp end parallel  
  ierr = 0_8
end subroutine
!........................................

!........................................
subroutine pusher_step5(particles, dt, b_part, kind_map, params_map) 

  implicit none
  real(kind=8), intent(inout)  :: particles (0:,0:)
  real(kind=8), value  :: dt
  real(kind=8), intent(in)  :: b_part (0:,0:)
  integer(kind=8), value  :: kind_map
  real(kind=8), intent(in)  :: params_map (0:)
  real(kind=8), allocatable  :: b (:) 
  real(kind=8), allocatable  :: b_prod (:,:) 
  real(kind=8), allocatable  :: v (:) 
  real(kind=8), allocatable  :: xi (:) 
  real(kind=8), allocatable  :: dfinv (:,:) 
  real(kind=8), allocatable  :: dfinv_t (:,:) 
  real(kind=8), allocatable  :: temp_mat1 (:,:) 
  real(kind=8), allocatable  :: temp_mat2 (:,:) 
  real(kind=8), allocatable  :: rhs (:) 
  real(kind=8), allocatable  :: identity (:,:) 
  real(kind=8), allocatable  :: lhs (:,:) 
  real(kind=8), allocatable  :: lhs1 (:,:) 
  real(kind=8), allocatable  :: lhs2 (:,:) 
  real(kind=8), allocatable  :: lhs3 (:,:) 
  integer(kind=8)  :: np  
  integer(kind=8), allocatable  :: components (:,:) 
  integer(kind=8)  :: ierr  
  integer(kind=8)  :: ip  
  real(kind=8)  :: det_lhs  
  real(kind=8)  :: det_lhs1  
  real(kind=8)  :: det_lhs2  
  real(kind=8)  :: det_lhs3  
  integer(kind=8)  :: i  
  integer(kind=8)  :: j  







  allocate(b(0:2_8))


  allocate(b_prod(0:2_8, 0:2_8))
  b_prod = 0.0


  allocate(v(0:2_8))
  allocate(xi(0:2_8))


  allocate(dfinv(0:2_8, 0:2_8))
  allocate(dfinv_t(0:2_8, 0:2_8))


  allocate(temp_mat1(0:2_8, 0:2_8))
  allocate(temp_mat2(0:2_8, 0:2_8))


  allocate(rhs(0:2_8))


  allocate(identity(0:2_8, 0:2_8))
  identity = 0.0
  identity(0_8, 0_8) = 1.0d0
  identity(1_8, 1_8) = 1.0d0
  identity(2_8, 2_8) = 1.0d0


  allocate(lhs(0:2_8, 0:2_8))


  allocate(lhs1(0:2_8, 0:2_8))
  allocate(lhs2(0:2_8, 0:2_8))
  allocate(lhs3(0:2_8, 0:2_8))


  np = size(particles(:, 0_8),1)


  allocate(components(0:2_8, 0:2_8))


  components(0_8, 0_8) = 11_8
  components(0_8, 1_8) = 12_8
  components(0_8, 2_8) = 13_8
  components(1_8, 0_8) = 21_8
  components(1_8, 1_8) = 22_8
  components(1_8, 2_8) = 23_8
  components(2_8, 0_8) = 31_8
  components(2_8, 1_8) = 32_8
  components(2_8, 2_8) = 33_8


  !$omp parallel
  !$omp do private(ip, b, xi, v, dfinv, dfinv_t, temp_mat1, temp_mat2, rhs, lhs,&
      !$omp & det_lhs, lhs1, lhs2, lhs3, det_lhs1, det_lhs2, det_lhs3) firstprivate(b&
      !$omp &_prod)
  do ip = 0, np - 1_8, 1


    b(0_8) = b_part(ip, 0_8)
    b(1_8) = b_part(ip, 1_8)
    b(2_8) = b_part(ip, 2_8)


    b_prod(0_8, 1_8) = -b(2_8)
    b_prod(0_8, 2_8) = b(1_8)


    b_prod(1_8, 0_8) = b(2_8)
    b_prod(1_8, 2_8) = -b(0_8)


    b_prod(2_8, 0_8) = -b(1_8)
    b_prod(2_8, 1_8) = b(0_8)


    xi = particles(ip, 0_8:2_8)
    v = particles(ip, 3_8:5_8)


    do i = 0, 2_8, 1
      do j = 0, 2_8, 1
        dfinv(i, j) = mapping_8z3wk9_df_inv(xi(0_8), xi(1_8), xi(2_8), &
      kind_map, params_map, components(i, j))


      end do

    end do

    call linalg_f7anys_transpose(dfinv, dfinv_t)


    call linalg_f7anys_matrix_matrix(b_prod, dfinv, temp_mat1)
    call linalg_f7anys_matrix_matrix(dfinv_t, temp_mat1, temp_mat2)


    call linalg_f7anys_matrix_vector((-dt)*(temp_mat2/Real(2_8, 8)) + &
      identity, v, rhs)


    lhs = dt*(temp_mat2/Real(2_8, 8)) + identity


    det_lhs = linalg_f7anys_det(lhs)


    lhs1(:, 0_8) = rhs
    lhs1(:, 1_8) = lhs(:, 1_8)
    lhs1(:, 2_8) = lhs(:, 2_8)


    lhs2(:, 0_8) = lhs(:, 0_8)
    lhs2(:, 1_8) = rhs
    lhs2(:, 2_8) = lhs(:, 2_8)


    lhs3(:, 0_8) = lhs(:, 0_8)
    lhs3(:, 1_8) = lhs(:, 1_8)
    lhs3(:, 2_8) = rhs


    det_lhs1 = linalg_f7anys_det(lhs1)
    det_lhs2 = linalg_f7anys_det(lhs2)
    det_lhs3 = linalg_f7anys_det(lhs3)


    particles(ip, 3_8) = det_lhs1/det_lhs
    particles(ip, 4_8) = det_lhs2/det_lhs
    particles(ip, 5_8) = det_lhs3/det_lhs




  end do

  !evaluate inverse Jacobian matrix
  !transpose of inverse Jacobian matrix
  !perform matrix-matrix and matrix-vector multiplications
  !explicit part of update rule
  !implicit part of update rule
  !solve 3 x 3 system with Cramer's rule
  !$omp end do  
  !$omp end parallel  
  ierr = 0_8
end subroutine
!........................................

end module

File: hylife/utilitis_PIC_April2020/STRUPHY_pusher.py
from pyccel.decorators import types
import ..linear_algebra.core as linalg
import ..geometry.mappings_analytical as mapping


# ==========================================================================================================
@types('double[:,:](order=F)','double','double[:,:](order=F)','double[:,:](order=F)','int','double[:]')
def pusher_step3(particles, dt, b_part, u_part, kind_map, params_map):
    
    from numpy import empty
    from numpy import zeros
    
    b          = empty( 3    , dtype=float)
    u          = empty( 3    , dtype=float)
    
    b_prod     = zeros((3, 3), dtype=float, order='F')
    
    xi         = empty( 3    , dtype=float)
    v          = empty( 3    , dtype=float)
    
    dfinv      = empty((3, 3), dtype=float, order='F')
    dfinv_t    = empty((3, 3), dtype=float, order='F')
    ginv       = empty((3, 3), dtype=float, order='F')
    
    temp_mat1  = empty((3, 3), dtype=float, order='F')
    temp_mat2  = empty((3, 3), dtype=float, order='F')
    
    temp_vec   = empty( 3    , dtype=float)
    
    np         = len(particles[:, 0])
    
    components = empty((3, 3), dtype=int, order='F')
    
    components[0, 0] = 11
    components[0, 1] = 12
    components[0, 2] = 13
    components[1, 0] = 21
    components[1, 1] = 22
    components[1, 2] = 23
    components[2, 0] = 31
    components[2, 1] = 32
    components[2, 2] = 33
    
    #$ omp parallel
    #$ omp do private (ip, b, u, xi, v, i, j, dfinv, dfinv_t, ginv, temp_mat1, temp_mat2, temp_vec) firstprivate(b_prod)
    for ip in range(np):
        
        b[0] = b_part[ip, 0]
        b[1] = b_part[ip, 1]
        b[2] = b_part[ip, 2]
        
        u[0] = u_part[ip, 0]
        u[1] = u_part[ip, 1]
        u[2] = u_part[ip, 2]
        
        b_prod[0, 1] = -b[2]
        b_prod[0, 2] =  b[1]

        b_prod[1, 0] =  b[2]
        b_prod[1, 2] = -b[0]

        b_prod[2, 0] = -b[1]
        b_prod[2, 1] =  b[0]
        
        xi = particles[ip, 0:3]
        v  = particles[ip, 3:6]
        
        # evaluate inverse Jacobian matrix
        for i in range(3):
            for j in range(3):
                dfinv[i, j] = mapping.df_inv(xi[0], xi[1], xi[2], kind_map, params_map, components[i, j])  
        
        # transpose of inverse Jacobian matrix
        linalg.transpose(dfinv, dfinv_t)
        
        # evaluate inverse metric tensor
        for i in range(3):
            for j in range(3):
                ginv[i, j] = mapping.g_inv(xi[0], xi[1], xi[2], kind_map, params_map, components[i, j]) 
                
        # perform matrix-matrix and matrix-vector products
        linalg.matrix_matrix(dfinv_t, b_prod, temp_mat1)
        linalg.matrix_matrix(temp_mat1, ginv, temp_mat2)
        linalg.matrix_vector(temp_mat2, u, temp_vec)
        
        # update particle velocities
        particles[ip, 3] += dt*temp_vec[0]
        particles[ip, 4] += dt*temp_vec[1]
        particles[ip, 5] += dt*temp_vec[2]
        
    #$ omp end do
    #$ omp end parallel 
        
    ierr = 0
    
    
    
# ==========================================================================================================
@external_call
@types('double[:,:](order=F)','double','int','double[:]')
def pusher_step4(particles, dt, kind_map, params_map):
    
    from numpy import empty
    
    xi         = empty( 3    , dtype=float)
    v          = empty( 3    , dtype=float)
    
    dfinv      = empty((3, 3), dtype=float, order='F')
    
    np         = len(particles[:, 0])
    
    components = empty((3, 3), dtype=int, order='F')
    
    components[0, 0] = 11
    components[0, 1] = 12
    components[0, 2] = 13
    components[1, 0] = 21
    components[1, 1] = 22
    components[1, 2] = 23
    components[2, 0] = 31
    components[2, 1] = 32
    components[2, 2] = 33
    
    k1 = empty( 3, dtype=float)  
    k2 = empty( 3, dtype=float)  
    k3 = empty( 3, dtype=float)  
    k4 = empty( 3, dtype=float)  
    
    #$ omp parallel
    #$ omp do private (ip, xi, v, dfinv, k1, k2, k3, k4)
    for ip in range(np):
        
        xi = particles[ip, 0:3]
        v  = particles[ip, 3:6]
        
        # step 1 in Runge-Kutta method
        for i in range(3):
            for j in range(3):
                dfinv[i, j] = mapping.df_inv(xi[0], xi[1], xi[2], kind_map, params_map, components[i, j])
                
        linalg.matrix_vector(dfinv, v, k1)
        
        # step 2 in Runge-Kutta method
        for i in range(3):
            for j in range(3):
                dfinv[i, j] = mapping.df_inv(xi[0] + dt*k1[0]/2, xi[1] + dt*k1[1]/2, xi[2] + dt*k1[2]/2, kind_map, params_map, components[i, j])
                
        linalg.matrix_vector(dfinv, v, k2)
        
        # step 3 in Runge-Kutta method
        for i in range(3):
            for j in range(3):
                dfinv[i, j] = mapping.df_inv(xi[0] + dt*k2[0]/2, xi[1] + dt*k2[1]/2, xi[2] + dt*k2[2]/2, kind_map, params_map, components[i, j])
                
        linalg.matrix_vector(dfinv, v, k3)
        
        # step 4 in Runge-Kutta method
        for i in range(3):
            for j in range(3):
                dfinv[i, j] = mapping.df_inv(xi[0] + dt*k3[0], xi[1] + dt*k3[1], xi[2] + dt*k3[2], kind_map, params_map, components[i, j])
                
        linalg.matrix_vector(dfinv, v, k4)
        
        # update logical coordinates
        particles[ip, 0] = (xi[0] + dt*(k1[0] + 2*k2[0] + 2*k3[0] + k4[0])/6)%1.
        particles[ip, 1] = (xi[1] + dt*(k1[1] + 2*k2[1] + 2*k3[1] + k4[1])/6)%1.
        particles[ip, 2] = (xi[2] + dt*(k1[2] + 2*k2[2] + 2*k3[2] + k4[2])/6)%1.
    
    #$ omp end do
    #$ omp end parallel 
        
    ierr = 0
    
    
    
# ==========================================================================================================
@types('double[:,:](order=F)','double','double[:,:](order=F)','int','double[:]')
def pusher_step5(particles, dt, b_part, kind_map, params_map):
    
    from numpy import empty
    from numpy import zeros
    
    b          = empty( 3    , dtype=float)
    
    b_prod     = zeros((3, 3), dtype=float, order='F')
    
    v          = empty( 3    , dtype=float)
    xi         = empty( 3    , dtype=float)
    
    dfinv      = empty((3, 3), dtype=float, order='F') 
    dfinv_t    = empty((3, 3), dtype=float, order='F')
    
    temp_mat1  = empty((3, 3), dtype=float, order='F')
    temp_mat2  = empty((3, 3), dtype=float, order='F')
    
    rhs        = empty( 3    , dtype=float)
    
    identity   = zeros((3, 3), dtype=float, order='F')
    identity[0, 0] = 1.
    identity[1, 1] = 1.
    identity[2, 2] = 1.
    
    lhs        = empty((3, 3), dtype=float, order='F')
    
    lhs1       = empty((3, 3), dtype=float, order='F')
    lhs2       = empty((3, 3), dtype=float, order='F')
    lhs3       = empty((3, 3), dtype=float, order='F')
    
    np         = len(particles[:, 0])
    
    components = empty((3, 3), dtype=int, order='F')
    
    components[0, 0] = 11
    components[0, 1] = 12
    components[0, 2] = 13
    components[1, 0] = 21
    components[1, 1] = 22
    components[1, 2] = 23
    components[2, 0] = 31
    components[2, 1] = 32
    components[2, 2] = 33
    
    #$ omp parallel
    #$ omp do private (ip, b, xi, v, dfinv, dfinv_t, temp_mat1, temp_mat2, rhs, lhs, det_lhs, lhs1, lhs2, lhs3, det_lhs1, det_lhs2, det_lhs3) firstprivate(b_prod)
    for ip in range(np):
        
        b[0] = b_part[ip, 0]
        b[1] = b_part[ip, 1]
        b[2] = b_part[ip, 2]
        
        b_prod[0, 1] = -b[2]
        b_prod[0, 2] =  b[1]
        
        b_prod[1, 0] =  b[2]
        b_prod[1, 2] = -b[0]
        
        b_prod[2, 0] = -b[1]
        b_prod[2, 1] =  b[0]
        
        xi = particles[ip, 0:3]
        v  = particles[ip, 3:6]
        
        # evaluate inverse Jacobian matrix
        for i in range(3):
            for j in range(3):
                dfinv[i, j] = mapping.df_inv(xi[0], xi[1], xi[2], kind_map, params_map, components[i, j])
        
        # transpose of inverse Jacobian matrix
        linalg.transpose(dfinv, dfinv_t)
        
        # perform matrix-matrix and matrix-vector multiplications
        linalg.matrix_matrix(b_prod, dfinv, temp_mat1)
        linalg.matrix_matrix(dfinv_t, temp_mat1, temp_mat2)
        
        # explicit part of update rule
        linalg.matrix_vector(identity - dt/2*temp_mat2, v, rhs)
        
        # implicit part of update rule
        lhs = identity + dt/2*temp_mat2
        
        # solve 3 x 3 system with Cramer's rule
        det_lhs = linalg.det(lhs)
        
        lhs1[:, 0] = rhs
        lhs1[:, 1] = lhs[:, 1]
        lhs1[:, 2] = lhs[:, 2]
        
        lhs2[:, 0] = lhs[:, 0]
        lhs2[:, 1] = rhs
        lhs2[:, 2] = lhs[:, 2]
        
        lhs3[:, 0] = lhs[:, 0]
        lhs3[:, 1] = lhs[:, 1]
        lhs3[:, 2] = rhs
        
        det_lhs1 = linalg.det(lhs1)
        det_lhs2 = linalg.det(lhs2)
        det_lhs3 = linalg.det(lhs3)
        
        particles[ip, 3] = det_lhs1/det_lhs
        particles[ip, 4] = det_lhs2/det_lhs
        particles[ip, 5] = det_lhs3/det_lhs
    
    #$ omp end do
    #$ omp end parallel     
        
    ierr = 0